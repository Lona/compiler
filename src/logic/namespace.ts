import { LogicAST as AST } from '@lona/serialization'
import NamespaceVisitor from './namespaceVisitor'

export type UUID = string

export type Namespace = {
  values: { [key: string]: UUID }
  types: { [key: string]: UUID }
}

export const builtInTypeConstructorNames: Set<String> = new Set([
  'Boolean',
  'Number',
  'String',
  'Array',
  'Color',
])

/**
 * Merge namespaces, throwing an error in the case of collisions
 */
export function mergeNamespaces(namespaces: Namespace[]): Namespace {
  return namespaces.reduce((result, namespace) => {
    Object.entries(namespace.values).forEach(([key, value]) => {
      if (key in result.values) {
        throw new Error(`Namespace error: value ${key} declared more than once`)
      }

      result.values[key] = value
    })

    Object.entries(namespace.types).forEach(([key, type]) => {
      if (key in result.types) {
        throw new Error(`Namespace error: type ${key} declared more than once`)
      }

      result.types[key] = type
    })

    return result
  }, createNamespace())
}

/**
 * Copy a namespace
 */
export function copy(namespace: Namespace): Namespace {
  return {
    values: { ...namespace.values },
    types: { ...namespace.types },
  }
}

/**
 * Build the global namespace by visiting each node.
 */
export function createNamespace(topLevelNode?: AST.SyntaxNode): Namespace {
  let namespace: Namespace = { types: {}, values: {} }

  if (topLevelNode) {
    let visitor = new NamespaceVisitor(namespace)

    visitor.traverse(topLevelNode)
  }

  return namespace
}
