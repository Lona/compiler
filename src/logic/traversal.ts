import { LogicAST as AST } from '@lona/serialization'
import { withOptions, IndexPath } from 'tree-visit'

const { visit, find, findAll } = withOptions({
  getChildren: AST.subNodes,
})

// Overload to support type guard
export function findNode<
  A extends AST.SyntaxNode['type'],
  T extends AST.SyntaxNode & { type: A }
>(
  rootNode: Parameters<typeof find>[0],
  predicate: (node: AST.SyntaxNode, indexPath: IndexPath) => node is T
): T | undefined

export function findNode(
  node: Parameters<typeof find>[0],
  predicate: (node: AST.SyntaxNode, indexPath: IndexPath) => boolean
): ReturnType<typeof find> {
  return find(node, predicate)
}

export const findNodes = findAll

export { visit }
