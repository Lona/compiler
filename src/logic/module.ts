import fs from 'fs'
import path from 'path'
import {
  decodeDocument,
  decodeLogic,
  extractProgramFromAST,
  LogicAST as AST,
  MDXAST,
} from '@lona/serialization'
import { IFS, match } from 'buffs'
import { evaluate, EvaluationContext } from './evaluation'
import { createNamespace, mergeNamespaces, Namespace } from './namespace'
import { createScopeContext, mergeScopes, Scope } from './scope'
import { createUnificationContext, TypeChecker } from './typeChecker'
import { Substitution, unify } from './typeUnifier'
import { joinPrograms, makeProgram } from './ast'

export type LogicFile = {
  isLibrary: boolean
  sourcePath: string
  rootNode: AST.TopLevelDeclarations
  mdxContent: MDXAST.Content[]
}

export type ModuleContext = {
  componentFiles: LogicFile[]
  libraryFiles: LogicFile[]
  documentFiles: LogicFile[]
  logicFiles: LogicFile[]
  sourceFiles: LogicFile[]
  namespace: Namespace
  scope: Scope
  typeChecker: TypeChecker
  substitution: Substitution
  evaluationContext: EvaluationContext
  isValidProgram: boolean
}

export function createModule(
  workspaceFs: IFS,
  workspacePath: string
): ModuleContext {
  const libraryFiles: LogicFile[] = libraryFilePaths().map(sourcePath => ({
    isLibrary: true,
    sourcePath,
    // Always read library files from the real FS
    rootNode: decodeLogic(
      fs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const componentFiles: LogicFile[] = componentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const logicFiles: LogicFile[] = logicFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => ({
    isLibrary: false,
    sourcePath,
    rootNode: decodeLogic(
      workspaceFs.readFileSync(sourcePath, 'utf8')
    ) as AST.TopLevelDeclarations,
    mdxContent: [],
  }))

  const documentFiles: LogicFile[] = documentFilePaths(
    workspaceFs,
    workspacePath
  ).map(sourcePath => {
    const decoded = decodeDocument(workspaceFs.readFileSync(sourcePath, 'utf8'))

    return {
      isLibrary: false,
      sourcePath,
      rootNode: extractProgramFromAST(decoded),
      mdxContent: decoded.children,
    }
  })

  const files: LogicFile[] = [
    ...libraryFiles,
    ...componentFiles,
    ...documentFiles,
    ...logicFiles,
  ]

  const namespace: Namespace = mergeNamespaces(
    files.map(logicFile => createNamespace(logicFile.rootNode))
  )

  const scope: Scope = mergeScopes(
    files.map(logicFile =>
      createScopeContext(logicFile.rootNode, namespace, undefined, console)
    )
  )

  const programNode = joinPrograms(
    files.map(logicFile => makeProgram(logicFile.rootNode))
  )

  const typeChecker = createUnificationContext(programNode, scope, console)

  const substitution = unify(typeChecker.constraints, console)

  const evaluationContext = evaluate(
    programNode,
    namespace,
    scope,
    typeChecker,
    substitution,
    console
  )

  return {
    componentFiles,
    libraryFiles,
    documentFiles,
    logicFiles,
    get sourceFiles() {
      return [...documentFiles, ...logicFiles]
    },
    namespace,
    scope,
    typeChecker,
    substitution,
    evaluationContext,
    isValidProgram:
      scope.undefinedIdentifierExpressions.size === 0 &&
      scope.undefinedMemberExpressions.size === 0 &&
      scope.undefinedTypeIdentifiers.size === 0,
  }
}

function componentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.cmp'] }).map(file =>
    path.join(workspacePath, file)
  )
}

function logicFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(workspacePath, file))
}

function documentFilePaths(fs: IFS, workspacePath: string): string[] {
  return match(fs, workspacePath, { includePatterns: ['**/*.md'] }).map(file =>
    path.join(workspacePath, file)
  )
}

export function libraryFilePaths(): string[] {
  return match(fs, STANDARD_LIBRARY_PATH, {
    includePatterns: ['**/*.logic'],
  }).map(file => path.join(STANDARD_LIBRARY_PATH, file))
}

export const STANDARD_LIBRARY_PATH = path.join(__dirname, '../../static/logic')
