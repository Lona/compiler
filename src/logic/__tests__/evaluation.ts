import { LogicAST as AST } from '@lona/serialization'
import { createFs } from 'buffs'
import { createModule, ModuleContext } from '../module'
import { UUID } from '../namespace'
import { findNode } from '../traversal'

function getInitializerId(module: ModuleContext, variableName: string): UUID {
  const logicFile = module.logicFiles.find(file =>
    file.sourcePath.endsWith('Example.logic')
  )

  if (!logicFile) {
    throw new Error(`Failed to find Example.logic`)
  }

  const variable = findNode(
    logicFile.rootNode,
    (node): node is AST.VariableDeclaration => {
      return node.type === 'variable' && node.data.name.name === variableName
    }
  )

  if (!variable) {
    throw new Error(`Variable ${variableName} not found`)
  }

  const initializer = variable.data.initializer

  if (!initializer) {
    throw new Error(`Initializer for ${variableName} not found`)
  }

  return initializer.data.id
}

function moduleWithFile(file: string): ModuleContext {
  const source = createFs({
    'lona.json': JSON.stringify({}),
    'Example.logic': file,
  })

  return createModule(source, '/')
}

describe('Logic / Evaluate', () => {
  it('evaluates number literals', () => {
    const file = `let x: Number = 4`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    expect(module.evaluationContext.evaluate(initializerId)).toEqual({
      type: { type: 'constructor', name: 'Number', parameters: [] },
      memory: { type: 'number', value: 4 },
    })
  })

  it('evaluates color literals', () => {
    const file = `let x: Color = #color(css: "red")`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    expect(module.evaluationContext.evaluate(initializerId)).toMatchSnapshot()
  })

  it('evaluates enums', () => {
    const file = `
enum Foo {
  case bar()
}

let x: Foo = Foo.bar()
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    expect(module.evaluationContext.evaluate(initializerId)).toEqual({
      type: { type: 'constructor', name: 'Foo', parameters: [] },
      memory: { type: 'enum', value: 'bar', data: [] },
    })
  })

  it('evaluates custom function', () => {
    const file = `
func test() -> Number {
  return 42
}

let x: Number = test()
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates custom function with arguments', () => {
    const file = `
func test(myNumber: Number) -> Number {
  return myNumber
}

let x: Number = test(myNumber: 42)
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates custom function with argument default value', () => {
    const file = `
func test(myNumber: Number = 42) -> Number {
  return myNumber
}

let x: Number = test()
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates DimensionSize', () => {
    const file = `
let x: DimensionSize = DimensionSize.fixed(100)
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    expect(module.evaluationContext.evaluate(initializerId)).toMatchSnapshot()
  })

  it('evaluates ElementParameter', () => {
    const file = `
let x: ElementParameter = ElementParameter.number("height", 20)
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates Element', () => {
    const file = `
let x: Element = Element(type: "Test", parameters: [])
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates Padding', () => {
    const file = `
let x: Padding = Padding(top: 10, right: 20, bottom: 30, left: 40)
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates function that returns Padding', () => {
    const file = `
let x: Padding = Padding.size(value: 8)
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  it('evaluates View', () => {
    const file = `
let width: DimensionSize = DimensionSize.fixed(20)
let height: DimensionSize = DimensionSize.fixed(20)
let name: String = "name"
let padding: Padding = Padding.size(value: 8)
let backgroundColor: Color = #color(css: "red")
let x: Element = View(__name: name, width: width, height: height, padding: padding, backgroundColor: backgroundColor, children: [])
`
    const module = moduleWithFile(file)
    const initializerId = getInitializerId(module, 'x')

    const result = module.evaluationContext.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })
})
