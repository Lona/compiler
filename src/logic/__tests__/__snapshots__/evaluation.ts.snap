// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic / Evaluate evaluates DimensionSize 1`] = `
Object {
  "memory": Object {
    "data": Array [
      Object {
        "memory": Object {
          "type": "number",
          "value": 100,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    ],
    "type": "enum",
    "value": "fixed",
  },
  "type": Object {
    "name": "DimensionSize",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates Element 1`] = `
Object {
  "memory": Object {
    "type": "record",
    "value": Object {
      "parameters": Object {
        "memory": Object {
          "type": "array",
          "value": Array [],
        },
        "type": Object {
          "name": "Array",
          "parameters": Array [
            Object {
              "type": "variable",
              "value": "?92",
            },
          ],
          "type": "constant",
        },
      },
      "type": Object {
        "memory": Object {
          "type": "string",
          "value": "Test",
        },
        "type": Object {
          "name": "String",
          "parameters": Array [],
          "type": "constant",
        },
      },
    },
  },
  "type": Object {
    "name": "Element",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates ElementParameter 1`] = `
Object {
  "memory": Object {
    "data": Array [
      Object {
        "memory": Object {
          "type": "string",
          "value": "height",
        },
        "type": Object {
          "name": "String",
          "parameters": Array [],
          "type": "constant",
        },
      },
      Object {
        "memory": Object {
          "type": "number",
          "value": 20,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    ],
    "type": "enum",
    "value": "number",
  },
  "type": Object {
    "name": "ElementParameter",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates Padding 1`] = `
Object {
  "memory": Object {
    "type": "record",
    "value": Object {
      "bottom": Object {
        "memory": Object {
          "type": "number",
          "value": 30,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "left": Object {
        "memory": Object {
          "type": "number",
          "value": 40,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "right": Object {
        "memory": Object {
          "type": "number",
          "value": 20,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "top": Object {
        "memory": Object {
          "type": "number",
          "value": 10,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    },
  },
  "type": Object {
    "name": "Padding",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates View 1`] = `
Object {
  "memory": Object {
    "type": "record",
    "value": Object {
      "parameters": Object {
        "memory": Object {
          "type": "array",
          "value": Array [
            Object {
              "memory": Object {
                "data": Array [
                  Object {
                    "memory": Object {
                      "type": "string",
                      "value": "backgroundColor",
                    },
                    "type": Object {
                      "name": "String",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                  Object {
                    "memory": Object {
                      "type": "record",
                      "value": Object {
                        "value": Object {
                          "memory": Object {
                            "type": "string",
                            "value": "red",
                          },
                          "type": Object {
                            "name": "String",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                      },
                    },
                    "type": Object {
                      "name": "Color",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                ],
                "type": "enum",
                "value": "color",
              },
              "type": Object {
                "name": "ElementParameter",
                "parameters": Array [],
                "type": "constant",
              },
            },
            Object {
              "memory": Object {
                "data": Array [
                  Object {
                    "memory": Object {
                      "type": "string",
                      "value": "children",
                    },
                    "type": Object {
                      "name": "String",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                  Object {
                    "memory": Object {
                      "type": "array",
                      "value": Array [],
                    },
                    "type": Object {
                      "name": "Array",
                      "parameters": Array [
                        Object {
                          "type": "variable",
                          "value": "?9b",
                        },
                      ],
                      "type": "constant",
                    },
                  },
                ],
                "type": "enum",
                "value": "elements",
              },
              "type": Object {
                "name": "ElementParameter",
                "parameters": Array [],
                "type": "constant",
              },
            },
            Object {
              "memory": Object {
                "data": Array [
                  Object {
                    "memory": Object {
                      "type": "string",
                      "value": "width",
                    },
                    "type": Object {
                      "name": "String",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                  Object {
                    "memory": Object {
                      "data": Array [
                        Object {
                          "memory": Object {
                            "type": "number",
                            "value": 20,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                      ],
                      "type": "enum",
                      "value": "fixed",
                    },
                    "type": Object {
                      "name": "DimensionSize",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                ],
                "type": "enum",
                "value": "dimension",
              },
              "type": Object {
                "name": "ElementParameter",
                "parameters": Array [],
                "type": "constant",
              },
            },
            Object {
              "memory": Object {
                "data": Array [
                  Object {
                    "memory": Object {
                      "type": "string",
                      "value": "height",
                    },
                    "type": Object {
                      "name": "String",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                  Object {
                    "memory": Object {
                      "data": Array [
                        Object {
                          "memory": Object {
                            "type": "number",
                            "value": 20,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                      ],
                      "type": "enum",
                      "value": "fixed",
                    },
                    "type": Object {
                      "name": "DimensionSize",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                ],
                "type": "enum",
                "value": "dimension",
              },
              "type": Object {
                "name": "ElementParameter",
                "parameters": Array [],
                "type": "constant",
              },
            },
            Object {
              "memory": Object {
                "data": Array [
                  Object {
                    "memory": Object {
                      "type": "string",
                      "value": "padding",
                    },
                    "type": Object {
                      "name": "String",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                  Object {
                    "memory": Object {
                      "type": "record",
                      "value": Object {
                        "bottom": Object {
                          "memory": Object {
                            "type": "number",
                            "value": 8,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                        "left": Object {
                          "memory": Object {
                            "type": "number",
                            "value": 8,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                        "right": Object {
                          "memory": Object {
                            "type": "number",
                            "value": 8,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                        "top": Object {
                          "memory": Object {
                            "type": "number",
                            "value": 8,
                          },
                          "type": Object {
                            "name": "Number",
                            "parameters": Array [],
                            "type": "constant",
                          },
                        },
                      },
                    },
                    "type": Object {
                      "name": "Padding",
                      "parameters": Array [],
                      "type": "constant",
                    },
                  },
                ],
                "type": "enum",
                "value": "padding",
              },
              "type": Object {
                "name": "ElementParameter",
                "parameters": Array [],
                "type": "constant",
              },
            },
          ],
        },
        "type": Object {
          "name": "Array",
          "parameters": Array [
            Object {
              "type": "variable",
              "value": "?1n",
            },
          ],
          "type": "constant",
        },
      },
      "type": Object {
        "memory": Object {
          "type": "string",
          "value": "LonaView",
        },
        "type": Object {
          "name": "String",
          "parameters": Array [],
          "type": "constant",
        },
      },
    },
  },
  "type": Object {
    "name": "Element",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates color literals 1`] = `
Object {
  "memory": Object {
    "type": "record",
    "value": Object {
      "value": Object {
        "memory": Object {
          "type": "string",
          "value": "red",
        },
        "type": Object {
          "name": "String",
          "parameters": Array [],
          "type": "constant",
        },
      },
    },
  },
  "type": Object {
    "name": "Color",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates custom function 1`] = `
Object {
  "memory": Object {
    "type": "number",
    "value": 42,
  },
  "type": Object {
    "name": "Number",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates custom function with argument default value 1`] = `
Object {
  "memory": Object {
    "type": "number",
    "value": 42,
  },
  "type": Object {
    "name": "Number",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates custom function with arguments 1`] = `
Object {
  "memory": Object {
    "type": "number",
    "value": 42,
  },
  "type": Object {
    "name": "Number",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates function that returns Padding 1`] = `
Object {
  "memory": Object {
    "type": "record",
    "value": Object {
      "bottom": Object {
        "memory": Object {
          "type": "number",
          "value": 8,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "left": Object {
        "memory": Object {
          "type": "number",
          "value": 8,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "right": Object {
        "memory": Object {
          "type": "number",
          "value": 8,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
      "top": Object {
        "memory": Object {
          "type": "number",
          "value": 8,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    },
  },
  "type": Object {
    "name": "Padding",
    "parameters": Array [],
    "type": "constant",
  },
}
`;
